### PR TITLE
Toddbell/unreal cloudscript

### PR DESCRIPTION
**Pull request by toddbellMSFT**
emit cloudscript after Unreal Tests complete. Refactors code to a shared class between the UCommandlet and the APfTestActor

### CODE REVIEW

File: targets/UnrealMarketplacePlugin/examplesource/ExampleProject/Source/ExampleProject/ACLoudScriptTestResultUploader.h.ejs [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+{
+    GENERATED_BODY()
+public:
+    void UploadToCloudscript(TArray<class UPlayFabTestContext*> SuiteTests);
+
+private:
+    void SuccessfulUpload(const PlayFab::ClientModels::FExecuteCloudScriptResult& result) const;
+
+    void UploadErrored(const PlayFab::FPlayFabCppError& error) const;
+};
```
Comment (MichaelMackendMS) [inline targets/UnrealMarketplacePlugin/examplesource/ExampleProject/Source/ExampleProject/ACLoudScriptTestResultUploader.h.ejs]:
```suggestion

    bool TestsAreComplete() const;

    bool ShouldQuitOnCompletion() const;
};
```

File: targets/UnrealMarketplacePlugin/examplesource/ExampleProject/Source/ExampleProject/ACloudScriptTestResultUploader.cpp.ejs [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        request,
+        PlayFab::UPlayFabClientAPI::FExecuteCloudScriptDelegate::CreateUObject(this, &ACloudScriptTestResultUploader::SuccessfulUpload),
+        PlayFab::FPlayFabErrorDelegate::CreateUObject(this, &ACloudScriptTestResultUploader::UploadErrored)
+    );
+}
+
+void ACloudScriptTestResultUploader::SuccessfulUpload(const PlayFab::ClientModels::FExecuteCloudScriptResult& result) const
+{
+    UE_LOG(LogTestResults, Log, TEXT("Cloud Upload Success: \n%s"), *result.toJSONString());
+}
```
Comment (MichaelMackendMS) [inline targets/UnrealMarketplacePlugin/examplesource/ExampleProject/Source/ExampleProject/ACloudScriptTestResultUploader.cpp.ejs]:
```suggestion
    if(TestsAreComplete() && ShouldQuitOnCompletion()) {
        FGenericPlatformMisc::RequestExit(true);
    }
}
```

File: targets/UnrealMarketplacePlugin/examplesource/ExampleProject/Source/ExampleProject/ACloudScriptTestResultUploader.cpp.ejs [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+void ACloudScriptTestResultUploader::UploadErrored(const PlayFab::FPlayFabCppError& error) const
+{
+    UE_LOG(LogTestResults, Error, TEXT("Cloud Upload Error: \n%s"), *error.GenerateErrorReport());
+}
```
Comment (MichaelMackendMS) [inline targets/UnrealMarketplacePlugin/examplesource/ExampleProject/Source/ExampleProject/ACloudScriptTestResultUploader.cpp.ejs]:
```suggestion
}

bool APfTestActor::TestsAreComplete() const {
    return SuiteState == PlayFabApiTestActiveState::COMPLETE;
}

bool APfTestActor::ShouldQuitOnCompletion() const {
    return GetWorld() && !GetWorld()->IsPlayInEditor();
}
```

File: targets/UnrealMarketplacePlugin/examplesource/ExampleProject/Source/ExampleProject/ACloudScriptTestResultUploader.cpp.ejs [hunk 4]
```diff
@@ -0,0 +1,76 @@
+<%- copyright %>
+
+#include "ACloudScriptTestResultUploader.h"
+
+#include "IPlayFab.h"
+#include "PlayFabCppBaseModel.h"
+#include "PlayFabError.h"
+#include "Core/PlayFabClientAPI.h"
+#include "PlayFabTestContext.h"
+
```
Comment (MichaelMackendMS) [inline targets/UnrealMarketplacePlugin/examplesource/ExampleProject/Source/ExampleProject/ACloudScriptTestResultUploader.cpp.ejs]:
```suggestion
#include "GenericPlatformMisc.h"
```

File: targets/UnrealMarketplacePlugin/examplesource/ExampleProject/Source/ExampleProject/ACLoudScriptTestResultUploader.h.ejs [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+#include "ACloudScriptTestResultUploader.generated.h"
+
+DECLARE_LOG_CATEGORY_EXTERN(LogTestResults, Log, All); 
+
+UCLASS()
+class ACloudScriptTestResultUploader : public AActor
+{
+    GENERATED_BODY()
+public:
+    void UploadToCloudscript(TArray<class UPlayFabTestContext*> SuiteTests);
```
Comment (MichaelMackendMS) [inline targets/UnrealMarketplacePlugin/examplesource/ExampleProject/Source/ExampleProject/ACLoudScriptTestResultUploader.h.ejs]: We sould probably not pass an array by value.

Comment (MichaelMackendMS) [inline targets/UnrealMarketplacePlugin/examplesource/ExampleProject/Source/ExampleProject/ACLoudScriptTestResultUploader.h.ejs]: I don't see anything inside the body of this function that should prevent it from being const.

**Comment by toddbellMSFT** [comment]
retest

**Pull request merged**